### PR TITLE
Various updates to old recommendations

### PR DESCRIPTION
**Pull request by peter-tackage**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,7 +19,7 @@ Lessons learned from Android developers in [Futurice](http://www.futurice.com).
 #### [Also keep dimens.xml DRY, define generic constants](#dimensxml)
 #### [Do not make a deep hierarchy of ViewGroups](#deephierarchy)
 #### [Avoid client-side processing for WebViews, and beware of leaks](#webviews)
-#### [Use Robolectric for unit tests, Robotium for connected (UI) tests, and AssertJ-Android for easier assertions in your Android tests](#test-frameworks)
+#### [Use JUnit for unit tests, Espresso for connected (UI) tests, and AssertJ-Android for easier assertions in your Android tests](#test-frameworks)
```
Comment (pedpess) [inline README.md]: What about Roboeletric for integration tests in order to test Android framework? L

File: README.md [hunk 2]
```diff
@@ -508,17 +507,19 @@ Read more at [Proguard](https://www.guardsquare.com/en/proguard/manual/examples)
 
 #### SharedPreferences
```
Comment (pedpess) [inline README.md]: Wouldn't be better to wrap that up in 1 topic? For instace, how to retain information. 1) Through intent, 2) Through SharedPreferences explaining that it stores key/value and just just and 3) ContentProviders, saying that if we want to have more complex data stored that would be an option but there is the drawback of writing too much boiler plate code.

File: README.md (hunk 1 above)
Comment (peter-tackage) [inline README.md]: Robolectric is typically a unit testing framework. It isn't suited for integration testing because you are testing against Robolectric, not Android (and the different versions of Android), so the results are not typically meaningful. In the majority of cases, I would just skip using Robolectric all together.

File: README.md (hunk 2 above)
Comment (peter-tackage) [inline README.md]: True. In my opinion, this section needs a rewrite because it doesn't acknowledge the middle ground between SharedPreferences and ContentProviders. ContentProviders have their place, but I'm not sure it makes sense to heavily recommend them when other, simpler options often suffice.

File: README.md [hunk 3]
```diff
@@ -146,22 +145,33 @@ Ultimately, be kind to other developers; don't force them to change their tool o
...
+**[Jackson](http://wiki.fasterxml.com/JacksonHome)** is a Java library for JSON serialization and deserialization, it has a wide-scoped and versatile API, supporting various ways of processing JSON: streaming, in-memory tree model, and traditional JSON-POJO data binding. 
+
+[Gson](https://code.google.com/p/google-gson/) is another popular choice and being a smaller library than Jackson, you might prefer it to avoid 65k methods limitation. Also, if you are using  
+
+[Moshi](https://github.com/square/moshi), another of [Square's](https://github.com/square) open source libraries, builds upon learnings from the development of Gson.
```
Comment (tomaszpolanski) [inline README.md]: Could also mention easy integration with [Kotlin](https://medium.com/square-corner-blog/kotlins-a-great-language-for-json-fcd6ef99256b)

File: README.md [hunk 4]
```diff
@@ -415,39 +424,29 @@ Even if you don't witness this explicitly in a layout file, it might end up happ
...
 
-**Use [Robolectric](http://robolectric.org/) only for unit tests, not for views.** It is a test framework seeking to provide tests "disconnected from device" for the sake of development speed, suitable specially for unit tests on models and view models. However, testing under Robolectric is inaccurate and incomplete regarding UI tests. You will have problems testing UI elements related to animations, dialogs, etc, and this will be complicated by the fact that you are "walking in the dark" (testing without seeing the screen being controlled).
+**Avoid [Robolectric](http://robolectric.org/)** Prior to the improved support for JUnit in the Android build system, Robolectric was promoted as a test framework seeking to provide tests "disconnected from device" for the sake of development speed. However, testing under Robolectric is inaccurate and incomplete as it works by providing mock implementations of the Android platform, which provides no guarantees of correctness. Instead, use a combination of JVM based unit tests and dedicated on-device integration tests.
```
Comment (tomaszpolanski) [inline README.md]: Small change of opinion :D

File: README.md (hunk 2 above)
Comment (peter-tackage) [inline README.md]: This should be done in a separate PR.

File: README.md (hunk 1 above)
Comment (peter-tackage) [inline README.md]: Actually, I have already done that.

**Comment by peter-tackage** [comment]
Ok, thanks for the review @tomaszpolanski and @pedpess. I'm going to merge this to keep things moving.
